src/tools/optimization-options.h

```diff
@@ -208,12 +208,29 @@ struct OptimizationOptions : public ToolOptions {
     if (debug) {
       passRunner.setDebug(true);
     }
-    for (auto& pass : passes) {
+    // If the last thing to do is inspect the module, do it after the
+    // pre-writing passes
+    // TODO: Make this less hacky?
+    int preWritingIndex = -1;
+    if (runningDefaultOptimizationPasses()) {
+      preWritingIndex = passes.size() - 1;
+      while (preWritingIndex > 0 &&
+             (passes[preWritingIndex] == "print" ||
+              passes[preWritingIndex] == "print-stack-ir" ||
+              passes[preWritingIndex] == "metrics")) {
+        --preWritingIndex;
+      }
+    }
+    for (int i = 0, size = passes.size(); i < size; ++i) {
+      auto& pass = passes[i];
       if (pass == DEFAULT_OPT_PASSES) {
         passRunner.addDefaultOptimizationPasses();
       } else {
         passRunner.add(pass);
       }
+      if (i == preWritingIndex) {
+        passRunner.addDefaultPreWritingPasses();
+      }
```

Comment:
Not very sure what the purpose of this code is. I guess what you intended was running `print`, `print-stack-ir`, and `metrics` only once at the end..? But this can make those run twice if a user specify those in command line.

Reply:
The purpose isn't to make them run only once, but to run them after Stack IR generation if they are at the end of the optimization pipeline. Any other pass at the end of the optimization pipeline should be run before Stack IR generation. So if a user specifies them twice, it is expected that they will run twice.